Lib/urllib/parse.py

```diff
@@ -1000,7 +1002,26 @@ def _splittype(url):
     match = _typeprog.match(url)
     if match:
         scheme, data = match.groups()
-        return scheme.lower(), data
+        scheme = scheme.lower()
+        if scheme in _schemes_disallowing_control_chars:
+            # Sanity check url data to avoid control characters.
+            #  https://bugs.python.org/issue14826
+            #  https://bugs.python.org/issue36276
+            # The same control characters check was adopted by Golang in:
+            #  https://go-review.googlesource.com/c/go/+/159157
+            # Isn't it odd to be performing validation within this utility
+            # function?  Yes... but it is in wide use in all of the right
+            # places where URLs need a sanity check to avoid potential security
+            # issues in newline delimited text based protocol implementations.
+            # This way many things get it for free without every use needing to
+            # be updated to explicitly sanity check the path contents.
+            global _control_char_re
+            if _control_char_re is None:
+                _control_char_re = re.compile('[\x00-\x1f\x7f-\x9f]')
+            if _control_char_re.search(data):
```

Comment:
`urlopen` encodes the url https://github.com/python/cpython/blob/750d74fac5c510e39958b3f79641fe54096ee54f/Lib/urllib/request.py#L1730 before calling `_splittype` below and regex match against encoded url might fail. Perhaps check against  `unquote(data)` as a safety check? 

Without `unquote` `urlopen("http://127.0.0.1:1234/?q=HTTP/1.1\r\nHeader: Value")` will not raise a `ValueError` and regex will try to match against `//127.0.0.1:1234/?q=HTTP/1.1%0D%0AHeader:%20Value` here instead of decoded value.

Reply:
Indeed, I noticed.  But realize that your proposed test cases use a _different_ url opening path than urllib.request.urlopen() calls.  We have too many of entry points.  but they all wind up in `http.client` where previous header injection fixes naturally went.

moving this there seems to work out.  our codepath that calls urllib.parse.quote() on the URL is fine, a quoted url is by definition sane and up for the server to handle.  so validating the quoted text will pass.  it's anything that allows raw unquoted urls in that is vulnerable.  such as urllib.request.urlopen() calls or direct http.client.HTTPConnection use.

updated.

also, agreed that we only need to check for ascii also per:

https://github.com/python/cpython/blob/750d74fac5c510e39958b3f79641fe54096ee54f/Lib/http/client.py#L1085